### PR TITLE
:warning: Marker Parsing and Structure

### PR DESCRIPTION
**Pull request by DirectXMan12**
This re-implements our parsing and generation logic in a more structured form.
I've tried to write up extended guides in the godoc for each new package, so check that
out until I put together a cohesive guide.

TODO:

- [x] a couple of documentation spots
- [ ] design doc overview
- [ ] better tests
- [ ] iron out last few backwards-incompatibilities
- [x] support type-level schema modifications

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @mengqiy this has both the comment extraction and marker parsing.

**Comment by DirectXMan12** [comment]
@redborian FYI

File: pkg/rbac/parser.go [hunk 1]
```diff
@@ -16,68 +16,81 @@ limitations under the License.
...
+			Kind:       "ClusterRole",
+			APIVersion: "rbac.authorization.k8s.io/v1",
+		},
+		Rules: rules,
+	}, "role.yaml"); err != nil {
+		return err
+	}
+
+	return nil
 }
```
Comment (droot) [inline pkg/rbac/parser.go]: RBAC Parser Reviewed. Looks good.

File: pkg/markers/parse.go [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+	case DescribesType:
+		return tryAnonLookup(name, r.forType)
+	case DescribesField:
+		return tryAnonLookup(name, r.forField)
+	default:
+		return nil
+	}
+}
+
+// tryAnonLookup handles tries looking up the given marker as both an struct-based
```
Comment (droot) [inline pkg/markers/parse.go]: typo: may be remove `handles`

File: pkg/genall/doc.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// Each Generator knows how to register its markers into a central Registry,
+// and then how to generate output using a Collector and some root packages.
+// Each generator can be considered to be the output type of a marker, for easy
+// command line parsing.
+//
+// Output and Input
+//
+// Generators output artifacts via an OutputRule.  OutputRules know how to
+// write output for different package-associated (code) files, as well as
+// condig files.  Each OutputRule should also be considered to be the output
```
Comment (droot) [inline pkg/genall/doc.go]: s/condig/coding

File: pkg/genall/genall.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	for _, gen := range g {
+		if err := gen.RegisterMarkers(reg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Generator knows how to register some set of markers, and then produce
+// output artificates based on loaded code containing those markers,
```
Comment (droot) [inline pkg/genall/genall.go]: s/artificates/artifacts

File: pkg/crd/spec.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+	if !hasStorage {
+		// just add the error to the first relevant package for this CRD,
+		// since there's no specific error location
+		packages[0].AddError(fmt.Errorf("CRD for %s has no storage version", groupKind))
+	}
+
+	p.CustomResourceDefinitions[groupKind] = crd
+}
```
Comment (droot) [inline pkg/crd/spec.go]: Reviewed crd parsing. This is bulk of the code I guess, looks pretty good on a high level.

File: cmd/controller-gen/main.go [hunk 6]
```diff
@@ -17,197 +17,321 @@ package main
...
+		Long: "Generate Kubernetes API extension resources and code.",
+		Example: `	# Generate RBAC manifests and crds for all types under apis/,
+	# outputting crds to /tmp/crds and everything else to stdout
+	controller-gen rbac crd paths=./apis/... output:crd:dir=/tmp/crds output:stdout
+
+	# Generate deepcopy implementations for a particular file
+	controller-gen deepcopy paths=./apis/v1beta1/some_types.go
+
+	# Run all the generators for a given project
+	controller-gen paths=./apis/...
```
Comment (droot) [inline cmd/controller-gen/main.go]: is `all` supported ? because that is the arg which is used by all the kubebuilder projects.

File: go.sum [hunk 7]
```diff
@@ -79,5 +113,6 @@ k8s.io/gengo v0.0.0-20190128074634-0689ccc1d7d6 h1:4s3/R4+OYYYUKptXPhZKjQ04WJ6Eh
 k8s.io/gengo v0.0.0-20190128074634-0689ccc1d7d6/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
 k8s.io/klog v0.2.0 h1:0ElL0OHzF3N+OhoJTL0uca20SxtYt4X4+bzHeqrB83c=
 k8s.io/klog v0.2.0/go.mod h1:Gq+BEi5rUBO/HRz0bTSXDUcqjScdoY3a9IHpCEIOOfk=
+sigs.k8s.io/controller-runtime v0.1.10 h1:amLOmcekVdnsD1uIpmgRqfTbQWJ2qxvQkcdeFhcotn4=
```
Comment (droot) [inline go.sum]: I remember removing this dependency. Are we reintroducing it ? or this is a mistake ?

**Comment by droot** [comment]
Have reviewed all the changes (except deepcopy-gen) and looks good overall, have minor comments. You can probably submit deepcopy-gen in a separate PR if you would like to get the other changes in.

File: go.sum (hunk 7 above)
Comment (DirectXMan12) [reply go.sum]: no, I think it's a mistake

**Comment by DirectXMan12** [comment]
Forgot webhook-gen.  We'll need to make sure webhook gen works with our new plans before release -- I stripped out the parts I thought we no longer needed, but I just want to make sure.

**Comment by DirectXMan12** [comment]
Added better tests for the low-level stuff.  High-level stuff and schema operations need tests.  We can follow up with those before release.  Let's try and get this part merged tomorrow.

File: cmd/controller-gen/main.go (hunk 6 above)
Comment (DirectXMan12) [reply cmd/controller-gen/main.go]: not specifying any generators means "all"

File: pkg/webhook/parser.go [hunk 8]
```diff
@@ -14,223 +14,155 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Package webhook contains libraries for generating webhookconfig manifests
+// from markers in Go source files.
+//
+// The markers take the form:
+//
+//  +kubebuilder:rbac:failurePolicy=<string>,groups=<[]string>,resources=<[]string>,verbs=<[]string>,versions=<[]string>,name=<string>,path=<string>,mutating=<bool>
```
Comment (droot) [inline pkg/webhook/parser.go]: typo: s/rbac/webhook ?

Comment (mengqiy) [reply pkg/webhook/parser.go]: @DirectXMan12 Please fix this.

Comment (mengqiy) [inline pkg/webhook/parser.go]: It seems it now requires all arguments to be in the same line. We need to remember to update the book.

File: pkg/markers/doc.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+// generally convenient and is the form used in many existing markers.
+//
+// Each of those argument types maps to the corresponding go type.  Pointers
+// mark optional fields (a struct tag, below, may also be used).  The empty
+// interface will match any type.
+//
+// Struct fields may optionally be annotated with the `marker` struct tag.  The
+// first argument is a name override.  If it's left blank (or the tag isn't
+// present), the camelCase version of the name will be used.  The only
+// additional argument defined is `optional`, which marks a field as optional
```
[the remark this answers is not in the document]
Comment (mengqiy) [reply pkg/markers/doc.go]: I misunderstood it. Solly has explained it to me f2f.

File: pkg/loader/loader.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+// Actual ASTs can be accessed via the Syntax field.
+func (p *Package) NeedSyntax() {
+	if p.Syntax != nil {
+		return
+	}
+	out := make([]*ast.File, len(p.CompiledGoFiles))
+	var wg sync.WaitGroup
+	wg.Add(len(p.CompiledGoFiles))
+	for i, filename := range p.CompiledGoFiles {
+		go func(i int, filename string) {
```
[the remark this answers is not in the document]
Comment (DirectXMan12) [reply pkg/loader/loader.go]: i is used by `out[i]`.  Need a closure to bind the variables from this loop iteration.

**Comment by DirectXMan12** [comment]
discussed with @mengqiy in person, will merge as per previous approval, then to master

**Pull request merged**